### ISSUE TITLE
Missing libyaml-devel on PCLinuxOS 64-bit

### ISSUE THREAD

**Issue by jzakiya** [rvm/rvm#3703]
Tried to install ruby-2.3.1 on PCLOS 64-bit and got this error:

```
rvm_error "There is no 'libyaml-devel' package available for installation and RVM does not know how to make that happen, please tell us how to install 'libyaml-devel' on 'PCLinuxOS-2016' here: https://github.com/rvm/rvm/issues
```

For PCLOS 64-bit the library is named:  "lib64yaml-devel"
This is true for all the 64-bit libraries which have 32-bit versions.

**Comment by pkuczynski** [comment]
This the same issue as in #3808

**Comment by pkuczynski** [comment]
Sorry #3709

**Comment by pkuczynski** [comment]
@jzakiya could you please test again after getting patched RVM: `rvm get branch /bugfix/3703-libyaml-pclos64`

**Comment by pkuczynski** [comment]
Could you also post output of following commands:

```
less /etc/redhat-release
dpkg --print-architecture
```

**Comment by jzakiya** [comment]
Here are all the outputs.
PCLinuxOS uses rpm packages, not debian.
-------------------------------------------------------------

[jzakiya@localhost ~]$ rvm get branch /bugfix/3703-libyaml-pclos64
Downloading https://get.rvm.io
Downloading
https://raw.githubusercontent.com/wayneeseguin/rvm/master/binscripts/rvm-installer.asc
Verifying /home/jzakiya/.rvm/archives/rvm-installer.asc
gpg: Signature made Wed 02 Nov 2016 03:59:26 PM EDT using RSA key ID
BF04FF17
gpg: Good signature from "Michal Papis (RVM signing) <mpapis@gmail.com>"
[unknown]
gpg: WARNING: This key is not certified with a trusted signature!
gpg:          There is no indication that the signature belongs to the
owner.
Primary key fingerprint: 409B 6B17 96C2 7546 2A17  0311 3804 BB82 D39D C0E3
     Subkey fingerprint: 62C9 E5F4 DA30 0D94 AC36  166B E206 C29F BF04 FF17
GPG verified '/home/jzakiya/.rvm/archives/rvm-installer'
Downloading
https://github.com/rvm/rvm/archive/bugfix/3703-libyaml-pclos64.tar.gz

Upgrading the RVM installation in /home/jzakiya/.rvm/
    RVM PATH line found in /home/jzakiya/.mkshrc /home/jzakiya/.profile
/home/jzakiya/.bashrc /home/jzakiya/.zshrc.
    RVM sourcing line found in /home/jzakiya/.profile
/home/jzakiya/.bash_profile /home/jzakiya/.zlogin.
Upgrade of RVM in /home/jzakiya/.rvm/ is complete.

# jzakiya,
#
#   Thank you for using RVM!
#   We sincerely hope that RVM helps to make your life easier and more
enjoyable!!!
#
# ~Wayne, Michal & team.

In case of problems: https://rvm.io/help and https://twitter.com/rvm_io

Upgrade Notes:

  * No new notes to display.

RVM reloaded!
[jzakiya@localhost ~]$ rvm -v
rvm 1.28.0 (bugfix/3703-libyaml-pclos64) by Michal Papis, Piotr Kuczynski,
Wayne E. Seguin [https://rvm.io/]
[jzakiya@localhost ~]$ rvm install ruby-2.3.3
Searching for binary rubies, this might take some time.


No binary rubies available for: pclinuxos/2017/x86_64/ruby-2.3.3.


Continuing with compilation. Please read 'rvm help mount' to get more
information on binary rubies.
Checking requirements for pclinuxos.
Installing requirements for pclinuxos.
Updating system..........
Error running 'requirements_pclinuxos_update_system ruby-2.3.3',
showing last 15 lines of
/home/jzakiya/.rvm/log/1483321454_ruby-2.3.3/update_system.log
++ [[ -z '' ]]
++ rvm_error 'There is no '\'''\'' package available for installation and
RVM does not know how to make that happen,
please tell us how to install '\'''\'' on '\''PCLinuxOS-2017'\'' here:
https://github.com/rvm/rvm/issues'
++ rvm_pretty_print stderr
++ case "${rvm_pretty_print_flag:=auto}" in
++ case "${TERM:-dumb}" in
++ case "$1" in
++ [[ -t 2 ]]
++ return 1
++ printf %b 'There is no '\'''\'' package available for installation and
RVM does not know how to make that happen,
please tell us how to install '\'''\'' on '\''PCLinuxOS-2017'\'' here:
https://github.com/rvm/rvm/issues\n'
There is no '' package available for installation and RVM does not know how
to make that happen,
please tell us how to install '' on 'PCLinuxOS-2017' here:
https://github.com/rvm/rvm/issues
++ return 49
++ return 49
Requirements installation failed with status: 49.
[jzakiya@localhost ~]$ less /etc/redhat-release -> PCLinuxOS release 2017
(PCLinuxOS) for x86_64
[jzakiya@localhost ~]$ dpkg --print-architecture
bash: dpkg: command not found
[jzakiya@localhost ~]$
-------------------------------------------------------------

**Comment by pkuczynski** [comment]
I made some progress but I would like to ask you to gist also: `cat /etc/os-release`

**Comment by pkuczynski** [comment]
@jzakiya have you had a chance to look at this? I am ready with new version just need the above info...

**Comment by jzakiya** [comment]
[jzakiya@localhost ~]$ cat /etc/os-release
cat: /etc/os-release: No such file or directory

**Comment by pkuczynski** [comment]
Could you also show me `uname -m`?

**Comment by jzakiya** [comment]
[jzakiya@localhost ~]$ uname -m
x86_64

**Comment by pkuczynski** [comment]
Perfect. @jzakiya I pushed some further changes. Could you please test again after getting latest `rvm get branch /bugfix/3703-libyaml-pclos64`. Thanks!

**Comment by jzakiya** [comment]
[jzakiya@localhost ~]$ rvm get branch /bugfix/3703-libyaml-pclos64
Downloading https://get.rvm.io
Downloading https://raw.githubusercontent.com/wayneeseguin/rvm/master/
binscripts/rvm-installer.asc
Verifying /home/jzakiya/.rvm/archives/rvm-installer.asc
gpg: Signature made Wed 02 Nov 2016 03:59:26 PM EDT using RSA key ID
BF04FF17
gpg: Good signature from "Michal Papis (RVM signing) <mpapis@gmail.com>"
[unknown]
gpg: WARNING: This key is not certified with a trusted signature!
gpg:          There is no indication that the signature belongs to the
owner.
Primary key fingerprint: 409B 6B17 96C2 7546 2A17  0311 3804 BB82 D39D C0E3
     Subkey fingerprint: 62C9 E5F4 DA30 0D94 AC36  166B E206 C29F BF04 FF17
GPG verified '/home/jzakiya/.rvm/archives/rvm-installer'
Downloading https://github.com/rvm/rvm/archive/bugfix/3703-libyaml-
pclos64.tar.gz

Upgrading the RVM installation in /home/jzakiya/.rvm/
    RVM PATH line found in /home/jzakiya/.mkshrc /home/jzakiya/.profile
/home/jzakiya/.bashrc /home/jzakiya/.zshrc.
    RVM sourcing line found in /home/jzakiya/.profile
/home/jzakiya/.bash_profile /home/jzakiya/.zlogin.
Upgrade of RVM in /home/jzakiya/.rvm/ is complete.

# jzakiya,
#
#   Thank you for using RVM!
#   We sincerely hope that RVM helps to make your life easier and more
enjoyable!!!
#
# ~Wayne, Michal & team.

In case of problems: https://rvm.io/help and https://twitter.com/rvm_io

Upgrade Notes:

  * No new notes to display.

RVM reloaded!


[jzakiya@localhost ~]$ rvm install ruby-2.4.0
Searching for binary rubies, this might take some time.
No binary rubies available for: pclinuxos/2017/x86_64/ruby-2.4.0.
Continuing with compilation. Please read 'rvm help mount' to get more
information on binary rubies.
Checking requirements for pclinuxos.
Installing requirements for pclinuxos.
Updating system.
Installing required packages: autoconf, lib64readline, lib64readline-devel,
lib64openssl-devel, automake, bison, libtool....
Error running 'requirements_pclinuxos_libs_install autoconf lib64readline
lib64readline-devel lib64openssl-devel automake bison libtool',
showing last 15 lines of
/home/jzakiya/.rvm/log/1483630741_ruby-2.4.0/package_install_autoconf_lib64readline_lib64readline-devel_lib64openssl-devel_automake_bison_libtool.log
++ for sbin_path in /sbin /usr/sbin /usr/local/sbin
++ [[ -d /usr/local/sbin ]]
++ [[
:/home/jzakiya/.rvm/gems/jruby-9.1.6.0/bin:/home/jzakiya/.rvm/gems/jruby-9.1.6.0@global/bin:/home/jzakiya/.rvm/rubies/jruby-9.1.6.0/bin:/usr/bin:/bin:/usr/local/bin:/usr/X11R6/bin/:/usr/games:/usr/lib/kde4/libexec:sbin:/usr/sbin:/usr/lib64/qt4/bin:/usr/lib64/qt5/bin:/usr/bin:/usr/sbin:/usr/lib64/kde4/libexec:/bin:/sbin:/usr/X11R6/bin:/usr/games:/usr/local/bin:/usr/local/sbin:/home/jzakiya/.rvm/bin:/home/jzakiya/bin:/home/jzakiya/.rvm/bin:/usr/bin:/usr/sbin:/usr/lib64/kde4/libexec:/bin:/sbin:/usr/X11R6/bin:/usr/games:/usr/local/bin:/usr/local/sbin:/home/jzakiya/.rvm/bin:/home/jzakiya/bin:/home/jzakiya/.rvm/bin:
!= *\:\/\u\s\r\/\l\o\c\a\l\/\s\b\i\n\:* ]]
++ [[ -n '' ]]
++ command_to_run=(__rvm_sudo -p "%p password required for '$*': "
"${command_to_run[@]}")
++ __rvm_sudo -p '%p password required for '\''synaptic install -y autoconf
lib64readline lib64readline-devel lib64openssl-devel automake bison
libtool'\'': ' synaptic install -y autoconf lib64readline
lib64readline-devel lib64openssl-devel automake bison libtool
++ [[ -p == \-\p ]]
++ printf '%p password required for '\''synaptic install -y autoconf
lib64readline lib64readline-devel lib64openssl-devel automake bison
libtool'\'': '
bash: printf: `p': invalid format character
++ shift 2
++ su -c 'synaptic install -y autoconf lib64readline lib64readline-devel
lib64openssl-devel automake bison libtool'
Password:
su: Authentication failure
++ return 1
++ return 1
Requirements installation failed with status: 1.
[jzakiya@localhost ~]$

**Comment by raelgc** [comment]
@jzakiya Just to double check: have you entered the correct password? Because we can read

> Password:
> su: Authentication failure

**Comment by jzakiya** [comment]
Before the last version, it got to a point where a window popped up asking
me to su login to install dependencies, but when I put my password in the
program crashed.
In the current version, gets to the line here

[jzakiya@localhost ~]$ rvm install ruby-2.4.0
Searching for binary rubies, this might take some
time.

No binary rubies available for:
pclinuxos/2017/x86_64/ruby-2.4.0.

Continuing with compilation. Please read 'rvm help mount' to get more
information on binary
rubies.

Checking requirements for pclinuxos.
Installing requirements for pclinuxos.
Updating system.
Installing required packages: autoconf, lib64readline, lib64readline-devel,
lib64openssl-devel, automake, bison, libtool...

it hangs, so I hit a return, and it crashed with the shown output.

Based on you question, I tried to install 2.4.0 again as before, and when
it got to the same point I put my password in
and a popup windows opens with the following message:

E: Command line option 'y' [from -y] is not known.

So apparently there's a problem with su logging in, and thus it can't begin
to install the dependencies it needs to continue.

Looking at the list of the dependencies, they are all already installed on
my system, so I guess it needs to know where they are.

**Comment by pkuczynski** [comment]
@jzakiya well it seems that the dependencies are finally recognized and there is an attempt to install them. Could be that not many people use RVM on PCLinuxOS and nobody reported the problem before. So let's keep up this work and we will find solution shortly, I am sure...

Right now the process fails on dependency installation. That's how we do it:

`synaptic install -y PACKAGE1 PACKAGE2 PACKAGE3...`

Is this correct way of installing packages on PCLinuxOS? If not, how the command should look like?

**Comment by pkuczynski** [comment]
The `-y` option is passed to enforce the installation instead of waiting for the user confirmation. If synaptic requires a different switch, what would that be?

**Comment by mpapis** [comment]
one of the problems could be with:
```bash
++ printf '%p password required for '\''synaptic install -y autoconf lib64readline lib64readline-devel lib64openssl-devel automake bison libtool'\'': ' bash: printf: `p': invalid format character
```
it shows error instead of the proper message - and the intent only shows only after the fail

**Comment by pkuczynski** [comment]
I had a bit or read about PCLinuxOS and apparently the base package manager is [apt-rpm](https://en.wikipedia.org/wiki/APT-RPM), so I am not sure why do we even use synaptic. @jzakiya could you please confirm? I would migrate all install commands to use apt-rpm...

**Comment by jzakiya** [comment]
Oh wow, I'm sorry, I thought you knew that.
Yes PCLOS uses apt-rpm to manage packages, synaptic I guess is just higher
level application sugar.

**Comment by pkuczynski** [comment]
I am not a PCLOS user so I do not know much about it :) Here are the commands which I figure out might work:

* Check if package installeld: `rpm -q --whatprovides "bzip2" >/dev/null || return $?`
* Check if package available: `apt-cache search "$1" >/dev/null 2>&1 || return $?`
* List package versions: `rpm -q --whatprovides "$1" | sed -e "s#^$1-##"`
* Install package: `apt-get -y install "$@"`

Let me know if these are the most sensible commands to run. Here is the commit: https://github.com/rvm/rvm/pull/3860/commits/cf366f6f3051cc098ee2edadc9f0a10afe426f4b

You might also try again `rvm get branch /bugfix/3703-libyaml-pclos64` and proceed with install to see if it works.

@jzakiya looking forward to hear from you soon!

**Comment by jzakiya** [comment]
You can use all the standard rpm commands|options that are in its
documentation.

Here is the output from the last rvm version.

----------------------------------------------------------------------------------------------------------

[jzakiya@localhost ~]$  rvm get branch /bugfix/3703-libyaml-pclos64
Downloading https://get.rvm.io
Downloading
https://bitbucket.org/mpapis/rvm/raw/master/binscripts/rvm-installer.asc
Verifying /home/jzakiya/.rvm/archives/rvm-installer.asc
gpg: Signature made Wed 02 Nov 2016 03:59:26 PM EDT using RSA key ID
BF04FF17
gpg: Good signature from "Michal Papis (RVM signing) <mpapis@gmail.com>"
[unknown]
gpg: WARNING: This key is not certified with a trusted signature!
gpg:          There is no indication that the signature belongs to the
owner.
Primary key fingerprint: 409B 6B17 96C2 7546 2A17  0311 3804 BB82 D39D C0E3
     Subkey fingerprint: 62C9 E5F4 DA30 0D94 AC36  166B E206 C29F BF04 FF17
GPG verified '/home/jzakiya/.rvm/archives/rvm-installer'
Downloading
https://github.com/rvm/rvm/archive/bugfix/3703-libyaml-pclos64.tar.gz
curl: (22) The requested URL returned error: 503 Service Unavailable

Could not download '
https://github.com/rvm/rvm/archive/bugfix/3703-libyaml-pclos64.tar.gz'.
  curl returned status '22'.

Downloading
https://bitbucket.org/mpapis/rvm/get/bugfix/3703-libyaml-pclos64.tar.gz
curl: (22) The requested URL returned error: 404 Not Found

Could not download '
https://bitbucket.org/mpapis/rvm/get/bugfix/3703-libyaml-pclos64.tar.gz'.
  curl returned status '22'.

Could not update RVM, get some help at #rvm IRC channel at freenode servers.

**Comment by pkuczynski** [comment]
Could you try again? Looks like temporal network issue...

**Comment by pkuczynski** [comment]
@jzakiya have you had a chance to try again?

**Comment by pkuczynski** [comment]
@jzakiya are you able to help with testing? Otherwise all the work will be wasted :(

**Comment by jzakiya** [comment]
Sorry, specifically what else would you like me to do?
If you tell me the explicit things you would like me to do then I can
follow them.

**Comment by pkuczynski** [comment]
Simple call:

```
rvm get branch /bugfix/3703-libyaml-pclos64
rvm install ruby-2.3.1
```

If that works we are done. If not, please gist the output.

**Comment by jzakiya** [comment]
Here is the gist of the output

https://gist.github.com/jzakiya/abfa71af35dc077b6fd53b0eee5eb9a7

When installing, is hangs on the line: Installing required packages:
autoconf, ......

I assumed it was waiting for my root password, so I just enter it, and it
continues on until failing.

**Comment by pkuczynski** [comment]
Ok, we definitely made a progress! Hopefully, we are close...

First of all you are right it was waiting for the password:

```
++ su -c 'apt-get -y install autoconf lib64readline lib64readline-devel lib64openssl-devel automake bison libtool'
Password: 
```

@mpapis how can we make it more explicit to the user he needs to provide the password? Ideas?

Then it fails with the missing package: `lib64readline`. I just updated the PR. Could you try again the same commands: 

```
rvm get branch /bugfix/3703-libyaml-pclos64
rvm install ruby-2.3.1
```

Thanks in advance!

**Comment by jzakiya** [comment]
Success!!

https://gist.github.com/jzakiya/9a3f7c2495307dfb5d60fd27cdbbf778

However, it still hangs at that line.
There needs to be a way to prompt for the password first to make it clear
to any user what is required.

I appreciate your persistence it getting it this far.

I thank you for all the PCLOS users.

**Comment by jzakiya** [comment]
I just installed ruby-2.4.0 too and all I had to do was

$ rvm install ruby-2.4.0

Since all the dependencies are now installed, it seems it will work with
all the MRI rubies.
Previously I couldn't install any RBX versions, will try the latest now.

**Comment by pkuczynski** [comment]
Success! Finally! :)

I will work on the password issue soon. Let me know if there were any problems with RBX

**Comment by jzakiya** [comment]
Installing rbx-3.70 failed, shown below.

https://gist.github.com/jzakiya/23e843d28cf4cc488f90fecc9528924f

**Comment by pkuczynski** [comment]
@jzakiya I suggest we raise a separate issue for RBX failing on PCLinuxOS so we can merge and close this one. Could you do that please?

**Comment by jzakiya** [comment]
Here's additional info on rbx

https://gist.github.com/jzakiya/fbb9dda8bacce173d2de88dff542c88f

**Comment by jzakiya** [comment]
What explicitly do you want me to do?

**Comment by pkuczynski** [comment]
I wanted you to create a separate issue for rbx. I did it myself to speed up things: #3895. We will continue working on rbx in that thread OK?

**Comment by pkuczynski** [comment]
The last thing to sort out here is the sudo password requirement. I prepared another patch. Could you please test it in following way:

```
apt-get -y remove lib64yaml-devel
rvm get branch /bugfix/3703-libyaml-pclos64
rvm install ruby-2.3.1
```

**Comment by pkuczynski** [comment]
By the way, does PCLinuxOS have `sudo`. In every other system we call `command sudo "$@"`, but on PCLinuxOS we use `su -c "$*"`. Maybe this hack is not necessary at all?

**Comment by jzakiya** [comment]
PCLOS uses "su" and not "sudo".

It is in its repos, but would have to be manually installed by a user.

Here is gist of successful installation of ruby-2.3.1.

https://gist.github.com/jzakiya/4bc3cfcf5ca71709f5412f77b2e28240

**Comment by pkuczynski** [comment]
@jzakiya could you try one more time:

```
apt-get -y remove lib64yaml-devel
rvm get branch /bugfix/3703-libyaml-pclos64
rvm install ruby-2.3.1
```

I would expect you to see a prompt for password.

**Comment by jzakiya** [comment]
It still hangs same place:


[jzakiya@localhost ~]$ rvm install ruby-2.3.1
Searching for binary rubies, this might take some time.
No binary rubies available for: pclinuxos/2017/x86_64/ruby-2.3.1.
Continuing with compilation. Please read 'rvm help mount' to get more
information on binary
rubies.

Checking requirements for pclinuxos.
Installing requirements for pclinuxos.
Updating system.
Installing required packages: lib64yaml-devel.....                 ###
hangs here until root password entered

**Comment by pkuczynski** [comment]
@jzakiya thanks for all your testing! Can you please run again and post the gist:

```
apt-get -y remove lib64yaml-devel
rvm get branch /bugfix/3703-libyaml-pclos64
rvm install ruby-2.3.1 --trace  2>&1 | tee rvm.log
```

I would expect you to see a prompt for password this time...

**Comment by jzakiya** [comment]
Yes, this created a proper password prompt.

The rvm.log file has 2,069,100 bytes and 16,487 lines.

https://gist.github.com/jzakiya/147f8646ffa48d706545465a5bd9c53f

I installed another ruby (2.2.6), which worked as expected.

Thank you for your work!

Is there anything else you need me to do?

Jabari

**Comment by pkuczynski** [comment]
How about now:

```
apt-get -y remove lib64yaml-devel
rvm get branch /bugfix/3703-libyaml-pclos64
rvm install ruby-2.3.1
```

Does this create proper password prompt? I would expected you to see a line:

```
[root] password required for 'apt-get -y install lib64yaml-devel'
```

**Comment by pkuczynski** [comment]
Hope this is one of the last things I ask in this ticket :)

**Comment by jzakiya** [comment]
Sorry not yet. :(

[jzakiya@localhost ~]$ rvm install ruby-2.3.1
Searching for binary rubies, this might take some time.
No binary rubies available for: pclinuxos/2017/x86_64/ruby-2.3.1.
Continuing with compilation. Please read 'rvm help mount' to get more
information on binary rubies.
Checking requirements for pclinuxos.
Installing requirements for pclinuxos.
Updating system.
Installing required packages: lib64yaml-devel....                  <----
still hangs with no message for root password

**Comment by mpapis** [comment]
I tracked down the problem to https://github.com/rvm/rvm/blob/master/scripts/functions/utility_logging#L181-L216 - especially lines 206 or 213 where we do `2>/dev/null`

we could save the stderr early before we call to `__rvm_log_dotted`, probably somewhere during initialization:
```bash
exec 6>&2
```
and then in `__rvm_setup_sudo_function_PCLinuxOS`:
```bash
printf "${2//%p/[root]/}" >&6
```

**Comment by mpapis** [comment]
@jzakiya I have added the proposed fix, can you check if it's asking for password now?

**Comment by jzakiya** [comment]
Ta da, dat did dit!  😂

[jzakiya@localhost ~]$ rvm install ruby-2.3.1
Searching for binary rubies, this might take some time.
No binary rubies available for: pclinuxos/2017/x86_64/ruby-2.3.1.
Continuing with compilation. Please read 'rvm help mount' to get more
information on binary rubies.
Checking requirements for pclinuxos.
Installing requirements for pclinuxos.
Updating system.
Installing required packages: lib64yaml-devel[root]/ password required for
'apt-get -y install lib64yaml-devel': .....

**Comment by jzakiya** [comment]
Here the gist of the whole process.

https://gist.github.com/jzakiya/2a6e1428523e219a23b829698a273579

**Comment by jzakiya** [comment]
I'm going to delete all those old gists on this issue on Monday, Feb 13,
2017, so if you need them please download them before tomorrow.
